design-proposals/generated-disk-serial.md

```diff
@@ -0,0 +1,202 @@
+# Overview
+This proposal introduces a new feature in KubeVirt to automatically
+generate serial numbers for each disk attached to a virtual machine.
+
+## Motivation
+- Windows guest OS in the VM requires that disk serial numbers are defined and persistent across reboots.
+- Generating the serial reduces manual configuration.
+
+## Goals
+- Add an option to the VM's `.spec` that will enable automatic serial number generation for disks that
+  don't have a serial number specified.
+
+- Generate unique serial numbers for each disk. The number has to be unique compared to all other disks attached
+  to all VMs with this disk.
+ 
+- Don't store the serial number in the VM's `.spec`. It would confuse gitops operators.
+
+- The serial number should remain the same across storage migration.
+
+## Non Goals
+
+- No customization options for serial number format.
+
+## Definition of Users
+
+- VM owners
+
+## User Stories
+
+- As a VM owner, I want to use the automated disk serial number assignment to reduce manual configuration.
+- As a VM owner, I want each disk to have a unique serial number.
+- As a VM owner, I want newly created Windows VMs to have serial numbers set on disks by default.
+
+## Repos
+- [kubevirt](https://github.com/kubevirt/kubevirt)
+
+# Design
+
+A new boolean option `generateDiskSerials` will be added to the VMI `.spec`. 
+When it is set to `true`, kubevirt will generate a serial number for each disk, that does not have one specified.
+These serial numbers will then be set in the libvirt domain xml.
+
+This option can also be added to `VirtualMachinePreference`, so it can be set on new VMs and
+the user does not need to be aware of it.
+
+## Persistence of generated serial numbers
+
+The serial number has to be the same across VM restarts and storage migration,
+so it has to be stored somewhere, or generated deterministically.
+
+This section lists several options how to store serial numbers.
+Then the rest of the document describes details of option: **Annotation on the PVC object**.
```

Comment:
Generally speaking the ID can be scoped in two ways:
- VM/I Internal - serial is only defined/kept within the VM/I bject
- VM/I external - serial is sourced from an external object - this is the case with the PVC annotation approach

If it's internal, then usually the requirements of a persistent serial number are met, however, it could lead to trouble if a volume is changed on a VMI, or if a volume is used on multiple VMIs.

If it is external, then we also meet the requirement of a persistent serial number, the same volume will get the same serial whenever attached to a VM that's good for vols attached to multi VMs, however, we need to think about cloning, and restore, and what it means for disk serial numbers. Something also requires permissionsto write the PVC.

Just to name a few things.